multi-educator/src/Dscp/MultiEducator/Launcher/Mode.hs

```diff
@@ -0,0 +1,176 @@
+import Dscp.Resource.Keys (BaseKeyParams (..), linkStore)
+import Dscp.Resource.Network
+import Dscp.Resource.SQLite (prepareEducatorSchema)
+import Dscp.Rio (RIO, runRIO)
+import Dscp.Util.HasLens
+import Dscp.Resource.Class (AllocResource, allocResource)
+import qualified Dscp.Witness as W
+
+---------------------------------------------------------------------
+-- WorkMode class
+---------------------------------------------------------------------
+
+-- | Set of typeclasses which define capabilities of bare Educator node.
+type MultiEducatorWorkMode ctx m =
+    ( Basic.BasicWorkMode m
+
+    , HasMultiEducatorConfig
+
+    , MonadReader ctx m
+
+    , HasLens' ctx (TVar EducatorContexts)
+
+    -- It's easier to just have these two lenses instead of reconstructing
+    -- the full educator context in multiToNormal from other lenses
+    , HasLens' ctx MultiEducatorResources
+    , HasLens' ctx W.WitnessContext
+
+    , MonadThrow m
+    )
+
+-- | Set of typeclasses which define capabilities both of Educator and Witness.
+type MultiCombinedWorkMode ctx m =
+    ( MultiEducatorWorkMode ctx m
+    , W.FullWitnessWorkMode ctx m
+    )
+
+---------------------------------------------------------------------
+-- WorkMode implementation
+---------------------------------------------------------------------
+
+data MultiEducatorContext = MultiEducatorContext
+    { _mecResources  :: !MultiEducatorResources
+      -- ^ Resources, allocated from params.
+    , _mecWitnessCtx :: !W.WitnessContext
+    }
+
+makeLenses ''MultiEducatorContext
+deriveHasLensDirect ''MultiEducatorContext
+
+type MultiEducatorRealMode = RIO MultiEducatorContext
+
+---------------------------------------------------------------------
+-- HasLens
+---------------------------------------------------------------------
+
+deriveHasLens 'mecResources ''MultiEducatorContext ''MultiEducatorResources
+deriveHasLens 'mecWitnessCtx ''MultiEducatorContext ''W.WitnessResources
+deriveHasLens 'mecWitnessCtx ''MultiEducatorContext ''W.WitnessVariables
+deriveHasLens 'mecWitnessCtx ''MultiEducatorContext ''NetServResources
+
+----------------------------------------------------------------------------
+-- (Almost) Natural Transformation
+----------------------------------------------------------------------------
+
+-- | This function transforms normal workmode into multi-workmode using login.
+-- It returns Nothing is login was not found
+lookupEducator :: MultiEducatorWorkMode ctx m => Text -> m (Maybe EducatorCtxWithCfg)
+lookupEducator login = do
+    mctx <- ask
+    let MultiEducatorResources{ _merEducatorData }
+            = mctx ^. lensOf @MultiEducatorResources
+    fmap (\(EducatorContexts ctxs) -> M.lookup login ctxs) . atomically $ readTVar _merEducatorData
+
+loadEducator :: MultiEducatorWorkMode ctx m => Bool -> Text -> Text -> m Bool
+loadEducator new login passphrase = do
+    -- TODO: add hashing
+    let loginFile = id (toString login)
```

Comment:
Current situation will allow login `./oxford` to access `oxford` account.

Reply:
... oh, and overwrite some `.key` files, too, if done carefully. Not mentioning crashing application when `dbsPath </> loginFile` is a directory or a protected file, like `$dbsPath/../../../protected.file`.

Reply:
Yes, all of the problems you've described are solved by hashing the login and using that as the file name (for both key and DB). At the time I was writing this code I didn't want to distract myself to do that and just left a `TODO` comment. In the final state of PR it will be there.